src/main/java/org/wildfly/extension/elytron/JdbcRealmDefinition.java

```diff
@@ -0,0 +1,529 @@
+        static final SimpleAttributeDefinition ALGORITHM = new SimpleAttributeDefinitionBuilder(ElytronDescriptionConstants.ALGORITHM, ModelType.STRING, false)
+                .setDefaultValue(new ModelNode(ScramDigestPassword.ALGORITHM_SCRAM_SHA_256))
+                .setValidator(new StringValuesValidator(
+                        ScramDigestPassword.ALGORITHM_SCRAM_SHA_1,
+                        ScramDigestPassword.ALGORITHM_SCRAM_SHA_256
+                ))
+                .setAllowExpression(false)
+                .setFlags(AttributeAccess.Flag.RESTART_RESOURCE_SERVICES)
+                .build();
+
+        static final SimpleAttributeDefinition PASSWORD = new SimpleAttributeDefinitionBuilder(ElytronDescriptionConstants.PASSWORD_INDEX, ModelType.INT, false)
+                .setMinSize(1)
+                .setAllowExpression(false)
+                .setFlags(AttributeAccess.Flag.RESTART_RESOURCE_SERVICES)
+                .build();
+
+        static final SimpleAttributeDefinition ITERATION_COUNT = new SimpleAttributeDefinitionBuilder(ElytronDescriptionConstants.ITERATION_COUNT_INDEX, ModelType.INT, false)
+                .setAllowExpression(false)
+                .setFlags(AttributeAccess.Flag.RESTART_RESOURCE_SERVICES)
+                .build();
+
+        static final SimpleAttributeDefinition SALT = new SimpleAttributeDefinitionBuilder(ElytronDescriptionConstants.SALT_INDEX, ModelType.INT, false)
+                .setAllowExpression(false)
+                .setFlags(AttributeAccess.Flag.RESTART_RESOURCE_SERVICES)
+                .build();
+
+        static final ObjectTypeAttributeDefinition SCRAM = new ObjectTypeAttributeDefinition.Builder(ElytronDescriptionConstants.SCRAM_MAPPER, ALGORITHM, PASSWORD, SALT, ITERATION_COUNT)
+                .setAllowNull(true)
+                .setFlags(AttributeAccess.Flag.RESTART_RESOURCE_SERVICES)
+                .build();
+
+        @Override
+        public ObjectTypeAttributeDefinition getResourceDefinition() {
+            return SCRAM;
+        }
+
+        @Override
+        public SimpleAttributeDefinition[] getAttributes() {
+            return new SimpleAttributeDefinition[] {ALGORITHM, PASSWORD, SALT, ITERATION_COUNT};
+        }
+
+        @Override
+        public PasswordKeyMapper toPasswordKeyMapper(OperationContext context, ModelNode propertyNode) throws OperationFailedException, InvalidKeyException {
+            String algorithm = ElytronExtension.asStringIfDefined(context, ALGORITHM, propertyNode);
+            int password = ElytronExtension.asIntIfDefined(context, PASSWORD, propertyNode);
+            int salt = ElytronExtension.asIntIfDefined(context, SALT, propertyNode);
+            int iterationCount = ElytronExtension.asIntIfDefined(context, ITERATION_COUNT, propertyNode);
+
+            return new PasswordKeyMapper(algorithm, password, salt, iterationCount);
+        }
+    }
+
+    interface PasswordMapperResourceDefinition {
```

Comment:
I think 'ResourceDefinition' is the wrong term to use here as that has a distinct definition already - maybe 'PasswordMapperObjectDefinition' and also rename the implementations?